Authoring Gradle Builds section in Gradle User Manual (#26992)
gradle/gradle: subprojects/docs/src/docs/userguide/authoring-builds/basics/writing_plugins.adoc

// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://creativecommons.org/licenses/by-nc-sa/4.0/
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

[[writing_plugins]]
= Writing Plugins

Writing custom plugins is the best way to customize and organize your build logic.

== Sample plugin

A plugin is any class that implements the link:{javadocPath}/org/gradle/api/Plugin.html[`Plugin`] interface.

The example below is the most straightforward plugin, a "hello world" *plugin*:

.build.gradle.kts
[source,kotlin]
----
import org.gradle.api.Plugin
import org.gradle.api.Project

abstract class SamplePlugin : Plugin<Project> {
    override fun apply(project: Project) {
        project.tasks.create("SampleTask") {
            println("Hello world!")
        }
    }
}
----

== Script plugin

> The simplest plugin is a script plugin coded in the build script.

Reviewer: 💭 A little weird to talk about the "simplest" plugin, when the previous example was simpler.  I think talking about subclassing `DefaultTask` is unnecessary, maybe this could use a built-in task?

    override fun apply(project: Project) {
        tasks {
            register("createFileTask", CreateFileTask::class) {
                group = "from my plugin"
                description = "Create myfile.txt in the current directory"
                fileText.set("HELLO FROM MY PLUGIN")
            }
        }
    }
}

apply<MyPlugin>()                                                                   // <3>
----
<1> Subclass `DefaultTask()`.
<2> Extend the `org.gradle.api.Plugin` interface.
<3> Apply the script plugin.

=== Subclass `DefaultTask()`
First, build a task by subclassing `DefaultTask()`.

[source]
----
abstract class CreateFileTask : DefaultTask() { }
----

This simple task adds a file to our application’s root directory.

=== Extend the `org.gradle.api.Plugin` interface
Next, create a new class that extends the `org.gradle.api.Plugin` interface.

[source]
----
abstract class MyPlugin : Plugin<Project> {
    override fun apply() {}
}
----

You can add tasks and other logic in the `apply()` method.

=== Apply the script plugin
Finally, apply the plugin in the build script.

[source]
----
apply<MyPlugin>()
----

> When `MyPlugin` is applied in the build script, it is calling the method `override fun apply() {}` as defined in the custom `MyPlugin` class.

Reviewer: ```suggestion
When `MyPlugin` is applied in the build script, Gradle calls the `fun apply() {}` method defined in the custom `MyPlugin` class.
```

.buildSrc/src/main/kotlin/MyConventionPlugin.kt
[source,kotlin]
----
import org.gradle.api.DefaultTask
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.provider.Property
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.OutputFile
import org.gradle.api.tasks.TaskAction
import java.io.File

abstract class CreateFileTask : DefaultTask() {
@get:Input
abstract val fileText: Property<String>

    @Input
    val fileName = project.rootDir.toString() + "myfile.txt"

    @OutputFile
    val myFile: File = File(fileName)

    @TaskAction
    fun action() {
        myFile.createNewFile()
        myFile.writeText(fileText.get())
    }
}

class MyConventionPlugin : Plugin<Project> {
    override fun apply(project: Project) {
        project.tasks.register("createFileTask", CreateFileTask::class.java) {
            group = "from my plugin"
            description = "Create myfile.txt in the current directory"
            fileText.set("HELLO FROM MY PLUGIN")
        }
    }
}
----

Which can then be applied in other build scripts accordingly:

.build.gradle.kts
[source,kotlin]
----
import com.gradle.plugin.MyConventionPlugin

> apply<MyConventionPlugin>()

Reviewer: 💭 This should probably show the newer `plugins {}` syntax.

Author: Can script plugins use the newer syntax?

Reviewer: Yes, as long as you define an ID for it.  Much of the existing docs use `apply` so that they can write the `Plugin` class and apply it to the project within a buildscript, for simplicity.  If you've already split the plugin out to it's own file in `buildSrc` it probably makes sense to show a complete example of defining the id and using the new syntax.  See https://dev.to/rogervinas/first-steps-developing-custom-gradle-plugins-4m51#create-plugins-in-the-buildsrc-module for an example.
